### PR TITLE
Timer

### PR DESCRIPTION
**Pull request by mwiktorin**

### CODE REVIEW

File: src/main/java/org/awesomebakery/TimeEngine.java [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+				}
+				synchronized (instance) {
+					currentDate.incrementOneHour();
+				}
+			}
+		}, ONE_MINUTE / TIME_FACTOR, ONE_MINUTE / TIME_FACTOR);
+
+	}
+
+	public synchronized static TimeEngine getInstance() {
```
Comment (sonarcube) [inline src/main/java/org/awesomebakery/TimeEngine.java]: MINOR Reorder the modifiers to comply with the Java Language Specification. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AModifiersOrderCheck)

File: src/main/java/org/awesomebakery/model/Date.java [hunk 2]
```diff
@@ -23,6 +23,30 @@ public int getHour() {
...
+
+	public boolean isBefore(Date date) {
+		if (day < date.day) {
+			return true;
+		}
+		return day == date.day && hour < date.hour;
+	}
+
+	@Override
+	public boolean equals(Object obj) {
```
Comment (sonarcube) [inline src/main/java/org/awesomebakery/model/Date.java]: MINOR This class overrides "equals()" and should therefore also override "hashCode()". [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1206)

File: src/main/java/org/awesomebakery/Start.java [hunk 3]
```diff
@@ -82,6 +82,10 @@ public static void main(String[] args) throws Throwable {
 
 		mainContainer.start();
 
+		TimeEngine.getInstance().registerPeriodicWakeUp(1, () -> {
```
Comment (sonarcube) [inline src/main/java/org/awesomebakery/Start.java]: MINOR Remove useless curly braces around statement [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1602)

**Comment by sonarcube** [comment]
SonarQube analysis reported 3 issues
* ![MINOR][MINOR] 3 minor

Watch the comments in this conversation to review them.

[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

File: src/main/java/org/awesomebakery/TimeEngine.java [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+	 * hour.
+	 */
+	public static final long TIME_FACTOR = 60;
+	private static final int ONE_MINUTE = 60 * 1000;
+
+	private static TimeEngine instance;
+
+	private Date currentDate;
+	private Map<Date, List<Runnable>> wakeUp;
+	private Timer timer;
```
Comment (element54) [inline src/main/java/org/awesomebakery/TimeEngine.java]: I don't like the idea to use a real timer. This way our simulation depends on real time.

File: src/main/java/org/awesomebakery/TimeEngine.java [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+		timer.scheduleAtFixedRate(new TimerTask() {
+			@Override
+			public void run() {
+				if (wakeUp.containsKey(currentDate)) {
+					for (Runnable runnable : wakeUp.get(currentDate)) {
+						runnable.run();
+					}
+				}
+				synchronized (instance) {
+					currentDate.incrementOneHour();
```
Comment (element54) [inline src/main/java/org/awesomebakery/TimeEngine.java]: Only increment by an our? We have periods that are in minutes (in future maybe even in seconds?)

**Pull request merged**